com.unity.render-pipelines.core/Runtime/XR/XRSystem.cs

```diff
@@ -0,0 +1,377 @@
+        #else
+            get => false;
+        #endif
+        }
+
+        /// <summary>
+        /// Valid empty pass when a camera is not using XR.
+        /// </summary>
+        public static readonly XRPass emptyPass = new XRPass();
+
+        /// <summary>
+        /// If true, the system will try to create a layout compatible with single-pass rendering.
+        /// </summary>
+        static public bool singlePassAllowed { get; set; } = true;
+
+        /// <summary>
+        /// If true, the system will log some information about the layout to the console.
+        /// </summary>
+        static public bool dumpDebugInfo { get; set; } = false;
+
+        /// <summary>
+        /// Use this method to assign the shaders that will be used to render occlusion mesh for each XRPass and the final mirror view.
+        /// </summary>
+        /// <param name="occlusionMeshPS"></param>
+        /// <param name="mirrorViewPS"></param>
+        public static void Initialize(Shader occlusionMeshPS, Shader mirrorViewPS)
+        {
+            RefreshDeviceInfo();
+
+            if (occlusionMeshPS != null)
+                s_OcclusionMeshMaterial = CoreUtils.CreateEngineMaterial(occlusionMeshPS);
+
+            if (mirrorViewPS != null)
+                s_MirrorViewMaterial = CoreUtils.CreateEngineMaterial(mirrorViewPS);
+
+            if (XRGraphicsAutomatedTests.enabled)
+                SetLayoutOverride(XRGraphicsAutomatedTests.OverrideLayout);
+        }
+
+        /// <summary>
+        /// Used by the render pipeline to communicate to the XR device how many samples are used by MSAA.
+        /// </summary>
+        /// <param name="msaaSamples"></param>
+        public static void SetDisplayMSAASamples(MSAASamples msaaSamples)
+        {
+            if (s_MSAASamples == msaaSamples)
+                return;
+
+            s_MSAASamples = msaaSamples;
+
+        #if ENABLE_VR && ENABLE_XR_MODULE
+            SubsystemManager.GetInstances(s_DisplayList);
+
+            foreach (var display in s_DisplayList)
+                display.SetMSAALevel((int)s_MSAASamples);
+        #endif
+        }
+
+        /// <summary>
+        /// Returns the number of samples (MSAA) currently configured on the XR device.
+        /// </summary>
+        /// <returns></returns>
+        public static MSAASamples GetDisplayMSAASamples()
+        {
+            return s_MSAASamples;
+        }
+
+        /// <summary>
+        /// Used by the render pipeline to scale the render target on the XR device.
+        /// </summary>
+        /// <param name="renderScale">A value of 1.0f represents 100% of the original resolution.</param>
+        public static void SetRenderScale(float renderScale)
+        {
+        #if ENABLE_VR && ENABLE_XR_MODULE
+            SubsystemManager.GetInstances(s_DisplayList);
+
+            foreach (var display in s_DisplayList)
+                display.scaleOfAllRenderTargets = renderScale;
+        #endif
+        }
```

Comment:
Scaling the render target for dynamic resolution should be avoided since XR must reallocate the eye texture whenever the scale changes. This has caused many developer performance issues before.
URP/Quest uses viewport scale and fixed foveated resolution for dynamic resolution. 
This refactor is a good time to drop SetRenderScale from URP and make it specific to HDRP if it still needs it.

RenderScale would be one more reason to create subclass from a base XRBaseSystem class to be used by each pipeline.

Reply:
This API method is currently called from each pipeline but we can remove it from URP if required. It's an optional call. I believe it's good to expose it publicly for the custom pipelines also. If URP no longer requires it, then we just don't call the method.